README
Perl-Critic-Compatibility version 0.999_001.

This distribution provides several Perl::Critic policies related to
compatibility with older versions of Perl.


INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install


NON-STANDARD BUILD TARGETS

    ./Build authortest

Runs extra tests that might not pass anywhere other than the author's computer.


COPYRIGHT AND LICENCE

Copyright (c)2008, Elliot Shank.  All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl-Critic-Compatibility

0.999_001  2008-04-??
    Initial release.


lib/Perl/Critic/Compatibility.pm
#      $URL: http://perlcritic.tigris.org/svn/perlcritic/trunk/Perl-Critic-Compatibility/lib/Perl/Critic/Compatibility.pm $
#     $Date: 2008-04-13 19:44:58 -0500 (Sun, 13 Apr 2008) $
#   $Author: clonezone $
# $Revision: 2230 $

package Perl::Critic::Compatibility;

use 5.006;

use strict;
use warnings;

our $VERSION = '0.999_001';

1; # Magic true value required at end of module

__END__

=for stopwords

=head1 NAME

Perl::Critic::Compatibility - Policies for Perl::Critic concerned with compatibility with older versions of Perl.

=head1 AFFILIATION

This module has no functionality, but instead contains documentation
for this distribution and acts as a means of pulling other modules
into a bundle.  All of the Policy modules contained herein will have
an "AFFILIATION" section announcing their participation in this
grouping.


=head1 VERSION

This document describes Perl::Critic::Compatibility version 1.0.0.


=head1 SYNOPSIS

Some L<Perl::Critic> policies that will help you keep your code
compatible with older versions of Perl, regardless of the version that
you're developing with.


=head1 DESCRIPTION

The included policies are:

=over

=item L<Perl::Critic::Policy::Compatibility::ProhibitThreeArgumentOpen>

Perls prior to 5.6 didn't support three argument open.  [Severity: 5]


=back


=head1 INTERFACE

None.  This is nothing but documentation.


=head1 DIAGNOSTICS

None.  This is nothing but documentation.


=head1 CONFIGURATION AND ENVIRONMENT

All policies included are in the "compatibility" theme.  See the
L<Perl::Critic> documentation for how to make use of this.


=head1 DEPENDENCIES

L<Perl::Critic>


=head1 INCOMPATIBILITIES

None reported.


=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-perl-critic-compatibility@rt.cpan.org>, or through the web
interface at L<http://rt.cpan.org>.


=head1 ACKNOWLEDGMENTS

Adam Kennedy for inspiring this.  Nay, demanding it.


=head1 AUTHOR

Elliot Shank  C<< <perl@galumph.com> >>


=head1 LICENSE AND COPYRIGHT

Copyright (c)2008, Elliot Shank C<< <perl@galumph.com> >>. All rights
reserved.

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT
WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER
PARTIES PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND,
EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE
SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME
THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE LIABLE
TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
DAMAGES.


=cut

# setup vim: set filetype=perl tabstop=4 softtabstop=4 expandtab :
# setup vim: set shiftwidth=4 shiftround textwidth=78 nowrap autoindent :
# setup vim: set foldmethod=indent foldlevel=0 :


lib/Perl/Critic/Policy/Compatibility/ProhibitThreeArgumentOpen.pm
#      $URL: http://perlcritic.tigris.org/svn/perlcritic/trunk/Perl-Critic-Compatibility/lib/Perl/Critic/Policy/Compatibility/ProhibitThreeArgumentOpen.pm $
#     $Date: 2008-04-13 19:44:58 -0500 (Sun, 13 Apr 2008) $
#   $Author: clonezone $
# $Revision: 2230 $

package Perl::Critic::Policy::Compatibility::ProhibitThreeArgumentOpen;

use 5.006;

use strict;
use warnings;

our $VERSION = '0.999_001';

use Readonly ();
use version ();

use Perl::Critic::Utils qw< :severities is_function_call parse_arg_list >;

use base 'Perl::Critic::Policy';

Readonly::Scalar my $DESCRIPTION => 'Three-argument form of open used';
Readonly::Scalar my $EXPLANATION => 'Three-argument open is not available until perl 5.6.';

Readonly::Scalar my $MINIMUM_VERSION => version->new(5.006);

sub supported_parameters { return ();                  }
sub default_severity     { return $SEVERITY_HIGHEST;   }
sub default_themes       { return qw( compatibility ); }
sub applies_to           { return 'PPI::Token::Word'   }

sub violates {
    my ( $self, $element, $document ) = @_;

    return if $element->content() ne 'open';
    return if not is_function_call($element);

    my $version = $document->highest_explicit_perl_version();
    return if $version && $version >= $MINIMUM_VERSION;

    my @arguments = parse_arg_list($element);

    if ( scalar @arguments > 2 ) {
        return $self->violation( $DESCRIPTION, $EXPLANATION, $element );
    }

    return;
} # end violates()

1;

__END__

=pod

=head1 NAME

Perl::Critic::Policy::Compatibility::ProhibitThreeArgumentOpen - Don't allow three-argument open unless the code uses a version of perl that supports it.


=head1 AFFILIATION

This policy is part of L<Perl::Critic::Compatibility>.


=head1 VERSION

This document describes
Perl::Critic::Policy::Compatibility::ProhibitThreeArgumentOpen version
1.000.


=head1 SYNOPSIS

Don't allow three-argument C<open> unless the module declares a
dependency upon perl 5.6 or higher.


=head1 DESCRIPTION

Perls prior to 5.6 don't support the three-argument form of C<open>.
If you want your code to remain compatible with those versions of
perl, you can't use it.

If your code explicitly declares a requirement on a version of perl
greater than or equal to 5.6, then three-argument C<open> is fine.

    open FILE, '<', 'blah';             # not ok

    use 5.006;
    open FILE, '<', 'blah';             # ok

    require 5.8.8;
    open FILE, '<', 'blah';             # ok


=head1 INTERFACE

Standard for a L<Perl::Critic::Policy>.


=head1 DIAGNOSTICS

None.


=head1 CONFIGURATION AND ENVIRONMENT

This policy has no configuration options beyond the standard ones.


=head1 DEPENDENCIES

L<Perl::Critic>


=head1 INCOMPATIBILITIES

None reported.


=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-perl-critic-compatible@rt.cpan.org>, or through the web
interface at L<http://rt.cpan.org>.


=head1 ACKNOWLEDGMENTS

Adam Kennedy for inspiring this.  Nay, demanding it.


=head1 AUTHOR

Elliot Shank  C<< <perl@galumph.com> >>


=head1 LICENSE AND COPYRIGHT

Copyright (c)2008, Elliot Shank C<< <perl@galumph.com> >>. All rights
reserved.

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT
WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER
PARTIES PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND,
EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE
SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME
THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE LIABLE
TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
DAMAGES.


=cut

# setup vim: set filetype=perl tabstop=4 softtabstop=4 expandtab :
# setup vim: set shiftwidth=4 shiftround textwidth=78 nowrap autoindent :
# setup vim: set foldmethod=indent foldlevel=0 :


Build.PL
#!perl

#      $URL: http://perlcritic.tigris.org/svn/perlcritic/trunk/Perl-Critic-Compatibility/Build.PL $
#     $Date: 2008-04-13 13:21:52 -0500 (Sun, 13 Apr 2008) $
#   $Author: clonezone $
# $Revision: 2221 $

use 5.006;

use strict;
use warnings;

use English qw< -no_match_vars >;

use Module::Build;


my $class = Module::Build->subclass( code => <<'END_SUBCLASS' );

    sub ACTION_authortest {
        my $self = shift;

        $self->depends_on('build');
        $self->depends_on('manifest');
        $self->depends_on('distmeta');
        $self->recursive_test_files(1);
        $self->depends_on('test');

        return;
    } # end ACTION_authortest()

    sub ACTION_distdir {
        my $self = shift;

        $self->depends_on('authortest');

        return $self->SUPER::ACTION_distdir();
    } # end ACTION_distdir

END_SUBCLASS

my $builder = $class->new(
    module_name     => 'Perl::Critic::Compatibility',
    license         => 'perl',
    dist_author     => 'Elliot Shank <perl@galumph.com>',
    dynamic_config  => 0,
    build_requires  => {
        'Test::More'    => 0,
    },
    requires => {
        'Perl::Critic'  => '1.083_001',
        'version'       => 0,
    },
    recommends      => {
    },
    create_packlist    => 1,
    sign               => 0,
    create_makefile_pl => 'traditional',
    add_to_cleanup     => [
        qw<
            MANIFEST
            MANIFEST.bak
            META.yml
            Makefile
            Makefile.PL
            pod2htm?.tmp
            Perl-Critic-Compatibility-*
        >
    ],
);


$builder->create_build_script();

# setup vim: set filetype=perl tabstop=4 softtabstop=4 expandtab :
# setup vim: set shiftwidth=4 shiftround textwidth=78 nowrap autoindent :
# setup vim: set foldmethod=indent foldlevel=0 :
